### ISSUE TITLE
some rules from crab tape recall are in state injected for several days

### ISSUE THREAD

**Issue by jhonatanamado** [dmwm/CMSRucio#420]
This issue was reported in the following [CMSTalk](https://cms-talk.web.cern.ch/t/crab-tape-recall-rucio-shows-inject-state/19693).
Digging what's going on with those rules in state injected I found that all of them comes from duplicates request. Somehow the Exception in [duplicate](https://github.com/dmwm/CRABServer/blob/ba11632d45e42512addcfaac681c1adbc1a5b870/src/python/TaskWorker/Actions/DBSDataDiscovery.py#L194-L205) rules is not working.
For this particular issue:

- I would like to understand how the `rse_expression` works. For those particular rules the pool of sites where the rule can be replicated is the same, but the string representation of the rse_expression is different.
- Also I noticed that those recalls are using `Async=True`. Could this somehow be involved that rucio "allows" the creation of the rules but never goes to the next state which is Replicating? the size of the request from crab is not huge. Why they are using `Async=True`
- One thing that I can not understand is that I couldn't create a rule for `/DYJetsToLL_M-50_Zpt-200toInf_BPSFilter_TuneCP5_13TeV-madgraphMLM-pythia8/RunIISummer20UL16RECOAPV-106X_mcRun2_asymptotic_preVFP_v8-v1/AODSIM` with the account `transfer_ops` against `T2_BE_UCL`. Returns an unique constraint error which is weird because Im using a different account.

**Comment by dciangot** [comment]
@jhonatanamado the taperecall rules still there? Do we have other example of this multiple injections? In other words, something that I can gather more information on. 

Regarding the last point, I see transfer_ops account already own a rule over that container with id 914383716b68400bb98255bae042d922 and destination T1_DE_KIT_Disk. I might be related.

**Comment by belforte** [comment]
let me rephrase the problem: it is surely possible that CRAB issues several tape recalls for same dataset, and as the comments in the code say, those should all be different rules since the top DID name is different (even if dataset/file content) is the same. And yes, we did not think this could be a problem (at worst a new set of locks on same files which will expire in 2 weeks anyhow, or hopefully only an expiration time update on existing lock). But then.. if the rule appears different to Rucio when it creates it, so that the DuplicateRule exception is not triggered, why did those rules get stuck in INJECT status ?

The 7 rules listed in the CMSTalk thread are all OK now. So not much to investigate.

We can add code to check if there's a rule already for same datasets in crab account, but I do not know if the queries involved will really result in less load for Rucio wrt some duplication.

**Comment by dciangot** [comment]
Personally, first of all, I'd try to find a log or a full understanding of these inject status means. So, unless there is some knowledge a priori on Eric/experts side, I'd try to catch some log for this kind of rules. Not easy, but it's worth a try probably

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
